src/librustc_trait_selection/traits/error_reporting/mod.rs

```diff
@@ -283,7 +283,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             .span_to_snippet(span)
                             .map(|s| &s == "?")
                             .unwrap_or(false);
-                        let is_from = format!("{}", trait_ref.print_only_trait_path())
+                        let is_from = format!("{}", trait_ref.print_only_trait_path_full())
                             .starts_with("std::convert::From<");
```

Comment:
String manipulation is really hacky, and code like this should instead check `trait_ref.def_id` for a lang item (if one exists) or a diagnostic item otherwise.

Reply:
Yeah, I agree it's hacky. But I think it should be a separate PR to tackle it. This PR is already a big change on its own.

Reply:
This predates the `lang_items` being in the codebase. We should change it as part of a separate PR.